server/blob/blob-deduplicating/src/main/scala/GC.scala

```diff
@@ -0,0 +1,192 @@
+                  blobsToDelete: Set[(Generation, BlobId)]
+                 )
+
+case class State(blobs: Set[BlobId],
+                 references: Map[Generation, Seq[Reference]],
+                 deletions: Map[Generation, Seq[Deletion]]) {
+
+  def stabilize(): StabilizedState = StabilizedState(references, deletions)
+
+  def apply(event: Event): State = { event match {
+    case e : Reference => copy(references = addElement(references, e), blobs = blobs + e.blob)
+    case e : Deletion => copy(deletions = addElement(deletions, e))
+  }
+  }
+
+  private def addElement[T <: Event](collection: Map[Generation, Seq[T]], e: T): Map[Generation, Seq[T]] = {
+    collection.updatedWith(e.generation)(previous => Some(e +: previous.getOrElse(Seq())))
+  }
+}
+
+case class StabilizedState(references: Map[Generation, Seq[Reference]],
+                           deletions: Map[Generation, Seq[Deletion]]) {
+  private val referencedBlobsAcrossGenerations: Map[Generation, ReferencedBlobs] = {
+    val maxGeneration = (references.keys ++ deletions.keys).maxOption.getOrElse(Generation.first)
+    val minGeneration = (references.keys ++ deletions.keys).minOption.getOrElse(Generation.first)
+
+    val initialRefs = Generation.range(InvalidGeneration, minGeneration.previous).map((_, ReferencedBlobs(Map()))).toMap
+    Generation.range(minGeneration, maxGeneration)
+      .foldLeft(initialRefs)(buildGeneration)
+  }
+
+  private def buildGeneration(refs: Map[Generation, ReferencedBlobs], generation: Generation): Map[Generation, ReferencedBlobs] = {
+    val populatedRefs = references.getOrElse(generation, Set())
+      .foldLeft(refs(generation.previous))((currentReferences, ref) => currentReferences.addReferences(ref.blobId))
+
+    val expungedRefs = deletions.getOrElse(generation, Set())
+      .foldLeft(populatedRefs)((currentReferences, ref) => currentReferences.removeReferences(ref.reference.blobId))
+
+    refs + (generation -> expungedRefs)
+  }
+
+  def referencesAt(generation: Generation): ReferencedBlobs = referencedBlobsAcrossGenerations(generation)
+
+  case class ReferencedBlobs(blobs: Map[BlobId, Int]) {
+    def isNotReferenced(blobId: BlobId): Boolean =
+      !blobs.contains(blobId)
+
+    def addReferences(blobId: BlobId): ReferencedBlobs =
+      ReferencedBlobs(blobs.updatedWith(blobId)(oldCount => oldCount.map(count => Some(count + 1)).getOrElse(Some(1))))
+    def removeReferences(blobId: BlobId): ReferencedBlobs =
+      ReferencedBlobs(blobs.updatedWith(blobId)(oldCount => oldCount.map(_ - 1).filter(_ > 0)))
+  }
+
+}
+
+object State {
+  val initial: State = State(blobs = Set.empty, references = Map.empty, deletions = Map.empty)
+}
+
+object Interpreter {
+  def apply(events: Seq[Event]): State =
+    events.foldLeft(State.initial)((state, event) => state(event))
+}
+
+object GC {
+  val temporization: Long = 2
+  def plan(state: StabilizedState, lastIteration: Iteration, targetedGeneration: Generation): Report = {
```

Comment:
I don't understand the arguments here.

Given a stabilized state, I just need to know the current generation, to which I apply the temporization, in order to get a report. 

(basically I don't know what the iteration argument is)

Also targetedGeneration sounds like a bad name to me (gc do perform the "switch" of generation?)

Reply:
* Generation: You should be able run a GC for a generation multiple times
 * Iteration: we need to know the set of Generations to run the GC on, so you need the last iteration

The GC increment the Iteration, the Generation is incremented externally.